### PR TITLE
Challenge sub trees and make them ready for payout

### PR DESCRIPTION
**Pull request by ayushmishra2005**
## Types of Changes
Please select the branch type you are merging and fill in the relevant template.

- [ ] Hotfix
- [ ] Release
- [ ] Fix or Feature

### Types of Changes

- [ ] Tech Debt (Code improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Dependency upgrade (A change in substrate or any 3rd party crate version)

### Migrations and Hooks

- [ ] This change requires a runtime migration.
- [ ] Modifies `on_initialize`
- [ ] Modifies `on_finalize`

### Checklist for Fix or Feature

- [ ] Change has been tested locally.
- [ ] Change adds / updates tests if applicable.
- [ ] Changelog doc updated.
- [ ] `spec_version` has been incremented.
- [ ] `network-relayer`'s [events](https://github.com/Cerebellum-Network/network-relayer/blob/dev-cere/shared/substrate/events.go) have been updated according to the blockchain events if applicable.
- [ ] All CI checks have been passed successfully

## Checklist for Hotfix

- [ ] Change has been deployed to Testnet.
- [ ] Change has been tested in Testnet.
- [ ] Changelog has been updated.
- [ ] Crate version has been updated.
- [ ] `spec_version` has been incremented.
- [ ] Transaction version has been updated if required.
- [ ] Pull Request to `dev` has been created.
- [ ] Pull Request to `staging` has been created.
- [ ] `network-relayer`'s [events](https://github.com/Cerebellum-Network/network-relayer/blob/dev-cere/shared/substrate/events.go) have been updated according to the blockchain events if applicable.
- [ ] All CI checks have been passed successfully

## Checklist for Release

- [ ] Change has been deployed to Devnet.
- [ ] Change has been tested in Devnet.
- [ ] Change has been deployed to Qanet.
- [ ] Change has been tested in Qanet.
- [ ] Change has been deployed to Testnet.
- [ ] Change has been tested in Testnet.
- [ ] Changelog has been updated.
- [ ] Crate version has been updated.
- [ ] Spec version has been updated.
- [ ] Transaction version has been updated if required.
- [ ] All CI checks have been passed successfully

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -30,6 +30,7 @@ resolver = "2"
 
 [workspace.dependencies]
 # 3rd-party dependencies
+base64 = { version = "0.22.1", default-features = false}
```
Comment (khssnv) [inline Cargo.toml]: Should we use "no_std" feature?

File: pallets/ddc-verification/src/lib.rs [hunk 2]
```diff
@@ -636,6 +666,98 @@ pub mod pallet {
...
+		pub record: Record,
+		pub transferred_bytes: u64,
+		pub stored_bytes: i64,
+	}
+
+	#[derive(
+		Debug, Serialize, Deserialize, Clone, Hash, Ord, PartialOrd, PartialEq, Eq, Encode, Decode,
+	)]
+	#[allow(non_snake_case)]
+	pub(crate) struct Record {
```
Comment (khssnv) [inline pallets/ddc-verification/src/lib.rs:717]: Would be better to replicate protobuf message version in a next version.

https://github.com/Cerebellum-Network/ddc-storage-node/blob/dev/pb/activity_record.proto

File: pallets/ddc-verification/src/lib.rs [hunk 3]
```diff
@@ -636,6 +666,98 @@ pub mod pallet {
...
+	)]
+	pub(crate) struct Upstream {
+		pub request: Request,
+	}
+
+	#[derive(
+		Debug, Serialize, Deserialize, Clone, Hash, Ord, PartialOrd, PartialEq, Eq, Encode, Decode,
+	)]
+	pub(crate) struct Downstream {
+		pub request: Request,
```
Comment (khssnv) [inline pallets/ddc-verification/src/lib.rs:736]: `request: Vec<Request>`?

File: pallets/ddc-verification/src/lib.rs [hunk 4]
```diff
@@ -636,6 +666,98 @@ pub mod pallet {
...
+		pub proofs: Vec<Proof>,
+	}
+
+	#[derive(
+		Debug, Serialize, Deserialize, Clone, Hash, Ord, PartialOrd, PartialEq, Eq, Encode, Decode,
+	)]
+	pub(crate) struct Proof {
+		pub merkle_tree_node_id: u64,
+		pub usage: Usage,
+		pub path: Vec<String>,
```
Comment (khssnv) [inline pallets/ddc-verification/src/lib.rs:686]: Future version may implement base64 deserialization in a trait.

File: pallets/ddc-verification/src/tests.rs [hunk 5]
```diff
@@ -2371,3 +2371,695 @@ fn test_bucket_node_aggregates() {
...
+		let host3 = "178.251.228.42";
+		let host4 = "37.27.30.47";
+
+		let port = 8080;
+
+		let pending_request1 = PendingRequest {
+			method: "GET".to_string(),
+			uri: format!("http://{}:{}/activity/buckets/123229/challenge?eraId=5757773&nodeId=0x1f50f1455f60f5774564233d321a116ca45ae3188b2200999445706d04839d72&merkleTreeNodeId=0,1,2,3", host1, port),
+			response: Some(br#"{
+  "proofs": [
```
Comment (khssnv) [inline pallets/ddc-verification/src/tests.rs:2440]: Would be nice to put those in a resource file.

File: pallets/ddc-verification/src/lib.rs (hunk 2 above)
Comment (khssnv) [reply pallets/ddc-verification/src/lib.rs:717]: I mean the next version of validation can define Record type similar to protobuf message. Also the access requested.

File: pallets/ddc-verification/src/lib.rs (hunk 3 above)
Comment (khssnv) [reply pallets/ddc-verification/src/lib.rs:736]: Ah yeah, sorry, I see we have `Vec<Downstream>` in the other place.

**Review by khssnv — APPROVED** [review]

File: pallets/ddc-verification/src/lib.rs [hunk 6]
```diff
@@ -1470,6 +1626,115 @@ pub mod pallet {
...
+						let leaf_record_hashes: Vec<ActivityHash> =
+							proof.leafs.into_iter().map(|p| p.hash::<T>()).collect();
+
+						let leaf_record_hashes_string: Vec<String> =
+							leaf_record_hashes.clone().into_iter().map(hex::encode).collect();
+
+						log::info!("🚀 Fetched leaf record hashes node id: {:?} bucket_id:{:?}  leaf_record_hashes: {:?}",
+						bucket_node_aggregate_activity.clone().node_id, bucket_id.clone(), leaf_record_hashes_string);
+
+						let leaf_record_root =
```
Comment (yahortsaryk) [inline pallets/ddc-verification/src/lib.rs:1676]: The root of an Aggregate Merkle tree must be calculated by taking Hash from the usage in the Root node (the node with `ID=1` based on the documentation). I don't believe you can calculate it only from tree leaves.

Correct me if I'm wrong:
cc: @khssnv  @upalinski @jwoloch

**Review by yahortsaryk — CHANGES_REQUESTED** [review]

File: pallets/ddc-verification/src/lib.rs [hunk 7]
```diff
@@ -705,6 +829,26 @@ pub mod pallet {
 		}
 	}
 
+	impl Activity for Leaf {
+		fn get_consensus_id<T: Config>(&self) -> ActivityHash {
+			T::ActivityHasher::hash(&self.record.id.encode()).into()
+		}
+
+		fn hash<T: Config>(&self) -> ActivityHash {
+			T::ActivityHasher::hash(&self.encode()).into()
```
[the remark this answers is not in the document]
Comment (jwoloch) [reply pallets/ddc-verification/src/lib.rs:838]: I confirm. Only the structures @yahortsaryk provided shall be hashed using scale encoding.

[the remark this answers is not in the document]
Comment (khssnv) [reply pallets/ddc-verification/src/lib.rs:838]: The hash is created from a leaf structure serialized with SCALE.

Here is the method which returns hasher input for a bucket activity tree: https://github.com/Cerebellum-Network/ddc-storage-node/blob/93e2d9d38e197f806c82589850f97c560a45b961/internal/activity/processing/tree.go#L44.

Here is the method which returns hasher input for a node activity tree: https://github.com/Cerebellum-Network/ddc-storage-node/blob/93e2d9d38e197f806c82589850f97c560a45b961/internal/activity/processing/tree.go#L95C10-L95C22.

Here we invoke an item serialization and pass it to the hasher: https://github.com/Cerebellum-Network/ddc-storage-node/blob/93e2d9d38e197f806c82589850f97c560a45b961/pkg/merkletree/activity/merkle_tree.go#L93-L100.

But it looks like we can skip this for now and fix in a future version.

File: pallets/ddc-verification/src/lib.rs (hunk 6 above)
Comment (jwoloch) [reply pallets/ddc-verification/src/lib.rs]: @yahortsaryk no, Hash of root is hash of child hashes.

Comment (yahortsaryk) [reply pallets/ddc-verification/src/lib.rs]: @jwoloch This is based on what @upalinski mentioned earlier
image

File: pallets/ddc-verification/src/lib.rs [hunk 8]
```diff
@@ -1470,6 +1631,115 @@ pub mod pallet {
...
+								cluster_id,
+								era_id,
+								&[merkle_root, path_hash],
+							)
+							.map_err(|err| vec![err])?;
+
+							log::info!("🚀 Fetched leaf node root node id: {:?} bucket_id:{:?} for path:{:?} leaf_node_hash: {:?}",
+						bucket_node_aggregate_activity.clone().node_id, bucket_id.clone(), path, hex::encode(node_root));
+
+							merkle_root = node_root;
```
Comment (yahortsaryk) [inline pallets/ddc-verification/src/lib.rs:1707]: Maybe I don't understand how exactly the `Self::create_merkle_root(...)` function works and in which circumstances it propagates an error, but I don't see where you are doing the exact *Assertion* of the fact that `leaves` returned in the `/challenge` response, really belong to the Merkle path (root hash) of the whole bucket sub-aggregate.

You are just iterating over `leafs` and `paths` from the response, creating sub-trees, but where is the exact comparison of the resulting hash of (`leaves` + `paths`) with the Merkle root hash (`hash` of the Root node with ID=`1`) ? Where does the exact challenge of the bucket sub-aggregate happen?

**Review by yahortsaryk — APPROVED** [review]

File: pallets/ddc-verification/src/lib.rs [hunk 9]
```diff
@@ -1470,6 +1633,203 @@ pub mod pallet {
...
+			let mut valid_aggregates_passed_challenges: Vec<BucketNodeAggregatesActivity> = vec![];
+
+			for (bucket_id, bucket_aggregates_passed_challenge_activities) in
+				bucket_aggregates_passed_challenges
+			{
+				let valid_activities = bucket_aggregates_passed_challenge_activities
+					.iter()
+					.cloned()
+					.max_by_key(|activity| activity.number_of_gets + activity.number_of_puts);
+
```
Comment (yahortsaryk) [inline pallets/ddc-verification/src/lib.rs:1747]: As a rule, the `number_of_gets` and `number_of_puts` should be the same between Aggregates provided by different Aggregators for the same Node or Bucket. This is the main topic of discussion we are having right now on how to compare deviating Aggregates and find the exact subtrees where discrepancies lie.

Also, 2 `put` requests for 2 MB has less economic value than 1 `put` request for 100 MB.

The `transferred_bytes` and `stored_bytes` is what we should refer to at first to define the economic "value" of the subaggegate. Please use a comparison by  total `transferred_bytes` / `stored_bytes` here

**Review by yahortsaryk — CHANGES_REQUESTED** [review]

**Review by yahortsaryk — APPROVED** [review]

**Pull request merged**